### ISSUE TITLE
OpenSSL can't see OPENSSL_MODULES env var when using Linux capabilities

### ISSUE THREAD

**Issue by ajbozarth** [openssl/openssl#21770]
### Background

OpenSSL, and specifically the libssl part of it, is primarily used for network encryption.
In Cloud, it is generally good practice that an application which uses of OpenSSL is NOT running as root. This prevents the application from binding to a privileged network:port (e.g. 443, typically any port < 1024), UNLESS the `net-cap-bind-service` capability is added to the application (e.g. `sudo setcap cap_net_bind_service=+ep /path/to/application`)
As a result, running an application in Cloud as non-root, but with addition of net-cap-bind-service capabilities, is in widespread use in Cloud.

### Issue

Linux capabilities provide a subset of the available root privileges to a process. Given the higher privileges, application programs with added capabilities are therefore started in a secure-execution mode in Linux. For security reasons, if the dynamic linker determines that a binary should be run in secure-execution mode, the effects of some environment variables are voided or modified, and furthermore ***those environment variables are stripped from the environment***, so that the program does not even see the definitions.

This means OpenSSL will not be able to see/read the `OPENSSL_MODULES` environment variable,  and therefore will never load a provider. This renders the providers useless in situations where capabilities are added to applications, specifically the case for net-cap-bind-service capability.

**Comment by ajbozarth** [comment]
@Planetf1 @mikebrow @psschwei @maximilien @johnma14 @martinschmatz this is what we discussed over Webex

**Comment by levitte** [comment]
As a fallback, there is a hard-coded directory, which is figured out with the command `openssl info -modulesdir`.  I can't say how you should run that command in your cloud environment, but if you can, that will tell you the default location to install the provider modules into.

**Comment by martinschmatz** [comment]
Agreed, but: 

The problems start when providers are not in this directory, and one can't copy the `provider.so` to that directory. The latter is the case when there aren't sufficient access rights to that directory, or when that directory doesn't exist anymore in the case one only uses the `library.so` and `provider.so` files for minimum size. 

The same btw is the case for the `openssl.cnf` file when using `OPENSSL_CONF` environment variable to specify this file. In that case, the hard coded `openssl version -d` will point to a potentially wrong location, which can lead to undetected use of a 'wrong' config file.

Last-not-least: A provider can offer configuration capabilities via environment variables, for example the `OQS OpenSSL provider` to adapt code points for interoperability.

**Comment by nhorman** [comment]
This seems like an administrative problem.  The os intentionally removes environment variables, so the solution would seem to me to be creating a shell wrapper around your applicaiton.  something like:
cat /path/to/application wrapper
#!/bin/sh
export OPENSSL_MODULES=/path/to/modules
exec /path/to/real/application

No?

**Comment by t8m** [comment]
The search path for providers can be set by application. Same for loading a configuration file. Of course the implicit loading of configuration file and providers would depend on the hardcoded path but there is no other way how this could be possibly set. Where it could get the paths from if not from application API call, hardcoded data, or environment variables?

**Comment by martinschmatz** [comment]
@t8m Reading from ENV variables is perfectly fine, IF those can be read. Which can be achieved by allowing reading ENV variables even when the net-cap-bind-service capability (and only this) is active. Similar to how `node.js` is doing it: See e.g. [here](https://github.com/nodejs/node/blob/c2c72601af8e17afde8f0faf3bf047d51ffb8b97/src/node_credentials.cc#L83) with helper functions [here](https://github.com/nodejs/node/blob/c2c72601af8e17afde8f0faf3bf047d51ffb8b97/src/node_credentials.cc#L52) and [here](https://github.com/nodejs/node/blob/c2c72601af8e17afde8f0faf3bf047d51ffb8b97/src/node_credentials.cc#L35) - functionality which could obviously be gated by build `define` statements. 
What can work for hard-coded file/directory paths in a cluster is to mount files to the 'right' position inside the container. But that would not address the situation where only ENV variables are available to configure a provider. And it would mean that the deployment process, not the (approved) image is controlling the content of the files. 

@nhorman I'll be happy to give the wrapper a try (after my 10 days vacation). I'll specifically will check whether the wrapper requires that no capabilities are added. If so, one would still have a process with elevated rights running - which could violate a policy defined by an administrative process to forbid exactly that.

**Comment by t8m** [comment]
OK, so you're asking for a concrete linux-specific ossl_safe_getenv() improvement - it would be perhaps better to say it from start. I think such improvement would be an acceptable PR if someone writes it. I am not seeing it as the OpenSSL team priority writing it right now (which does not mean it cannot change).

**Comment by martinschmatz** [comment]
@t8m Many thanks for considering! 
Weren't sure whether a readily available solution was overlooked. 

A PR submission sounds good to me. We'll check how/when that can be done.

**Comment by paulidale** [comment]
It's not at all clear to me what is being asked for here.
The OS/loader is stripping environment variables.
I don't see how they can be restored reasonably.
I.e. I don't see why this would be a problem with OpenSSL.

Modifying the application to set the various paths via OpenSSL calls or calling `putenv(3)` or `setenv(3)` seems like the sensible fix.

**Comment by t8m** [comment]
If I understand the problem right, it is not a problem that the environment variables aren't set. The problem is that the ossl_safe_getenv() treats the process as one with elevated privileges and thus disregards them.

**Comment by ajbozarth** [comment]
> The problem is that the ossl_safe_getenv() treats the process as one with elevated privileges and thus disregards them.

This is the problem as we see it.

We are currently working on a draft PR to show our potential solution for further discussion.

**Comment by t8m** [comment]
IMO the proposal is OK apart from some minor details. I do not think we need any other tests than a new build added to CI (probably the runchecker CI) that enables this code to be built. As executing with this extra capability bit set would be a problem.

**Comment by nhorman** [comment]
@t8m you're correct, the problem is that glibc is disregarding set environment variables, but its doing for a good reason.  Because the application has elevated privs (specifically, CAP_NET_BIND_SERVICE), secure_getenv doesn't allow the set env vars to be seen, reasoning that, with elevated privs (even just CAP_NET_BIND_SERVICE), that the risk of fault injection through those variables is risky.  With this change, we're effectively establishing the policy that CAP_NET_BIND_SERVICE is actually safe and we can go get those env vars from the parent shell.

My argument is that we shouldn't be setting that policy, at least not in a hard coded fashion, because we don't know what an end user would consider 'safe', nor do we know what capabilities an end user might set.  If we want a mechanism that allows environment variables to be considered safe to use, it should be done via an api that allows users to specify those conditions (i.e. what variables, and with which capabilities).  Hard coding the capabilieis and env vars to pass through invites follow on issues, in which subsequent users ask for env x with capability set y, that we have to review on a case by case basis.

I think its better to have an api that allows for that specification.

Alternatively, As mentioned above, this is all circumventable with a wrapper script that specifically re-introduces the needed env vars once established in the process running with elevated privs.

**Comment by t8m** [comment]
The problem with such API is that the application might not be the right place to call that API. It might not even know whether such API call should be needed. So yeah, I am afraid we are are back at the wrapper script thing.

**Comment by t8m** [comment]
> For security-out-of-the-box, it can be preferable to use a custom build of the OpenSSL libraries, in particular to disable lower SSL/TLS modes of operation already during build time and/or for tight version control and/or to have custom library symbols and/or ... . Therefore, gating the proposed functionality via build defines is - at least for me - not an issue (but might be for other folks).

If this is supposedly a custom build of OpenSSL then it might very well use different handling for the defaults applied to the various directories. We might for example make a build time option to force all these directories to be the directory where the library or executable is loaded from.

**Comment by martinschmatz** [comment]
Would be helpful IF the people building the OpenSSL/provider libs would be the same that consume the libs. 

If not, then using the environment variables is the way to eliminate the - potentially error prone - communication requirement between them. 
Just to mention: The folks deploying the image are again different persons who do not know per-se anything about environment variables or their content (read: without yet another communication effort). 

Nevertheless, unrelated to the current feature request, the idea to have control over the hard-coded default directories during _build time_ looks like a good idea to me. This would be the build time equivalent to the ability to control the location of the directories via environment variables during _run time_.

Pls don't get me wrong: I'm not saying that the problem could not be solved with some effort in some other ways. But the lean method IMHO is to use the environment variables during image build - which I guess is why they exist in the first place.

**Comment by nhorman** [comment]
In a containerized environment, isn't this problem almost trivial to solve? That is to say, the defined container image should be able to specify the needed environment variables and capabilities at the same time.  I would have thought that docker (or whatever container runtime is in use) would apply environment variables after setting capabilities.  Otherwise, a container would never receive any environment variables with any capabilities set (i.e. secure_getenv would never work in that environment).  If thats not the case, then it would seem like a bug in the container runtime.

**Comment by martinschmatz** [comment]
@nhorman (also see above): 

(1) Yes, it can be addressed at **_launch time_**, e.g. in a deployment `yaml` for k8s to provide access to environment variable `POD_NAME` despite setting `net_bind_cap` capabilities:
```
          securityContext:
            capabilities:
              drop:
                - ALL
              add:
                - NET_BIND_SERVICE
            runAsUser:  1000
            runAsGroup: 1000
          allowPrivilegeEscalation: false
          env:
            - name: POD_NAME
              valueFrom:
                fieldRef:
                  fieldPath: metadata.name
```

(2) No, when trying to address it at (image-)**_build time_**, I can add `ENV` variables in the Dockerfile before or after setting capabilities and/or before or after switching to `USER 1000` and never got an app-specific `openssl.cnf` in a custom location loaded. 

Therefore: When a team building an application image has an app-specific, policy-driven `openssl.cnf` in a place different than the OpenSSL lib build-time-hard-coded location, and duly defines the `OPENSSL_CONF` variable accordingly in the Dockerfile and also applies (= limits) capabilities to the application to prevent damage if the app would go ballistic, BUT the team deploying the application image by mistake/miscommunication/stupidity does not specify the `OPENSSL_CONF` variable (and/or the `OPENSSL_MODULES` env variable), the _default_ `openssl.cnf` provided by the team building the OpenSSL libs is loaded - which is super difficult to notice because the application might work, but might allow features which the application development team wanted to forbid! This breaks what I call "security out of the box".

I'd agree that one could argue that the deployment team would have to be more careful in such situations, but - without proof - human error is more likely to generate security risks than anything else.

**Comment by nhorman** [comment]
So to re-iterate what (I think) you're saying:
You have two distinct teams:
   a) container builders
   b) container deployers


Your builder team further specifies environment variables to a policy-specific configuration and points OPENSSL_CONF/OPENSSL_MODULES to the appropriate location to drive that policy

Your deployment team then deploys this container with said policy, but _if_ they fail to include corresponding environment variables to point OPENSSL_CONF/OPENSSL_MODULES to the same location that is specified at b the build team in the docker file, Those build time environment variables will get ignored (due to the use of secure_getenv stripping them), the defaults will instead be used in the container, leading to potential policy-violations from the use of the non-policy specified configuration.

Please correct me if I have any of that incorrect.

Reading that (assuming I've done so correctly), I would think that is a k8s/openshift/container environment bug.  I say that because, in the situation you describe, no use of secure_getenv would be possible (in openssl or elsewhere).  That is to say, any application that is built into a container that relies on both an environment variable and requires elevated privileges while running as a non-root user will be impacted by this situation, and as such should arguably be fixed by the container management environment.

Reading through some of the openshift docs, I get the impression that no one has really thought about this, but I would imagine that, given that a container management environment has no insight into the container applications' use of getenv vs. secure_getenv, they would be hard pressed to come up with a reasonable fix for it.

That said, If your goal is to provide "security out of the box", it would seem that the onus to do so is on the build team, and based on my prior arguments, I don't think allowing environment variables that have been stripped by glibc via secure_getenv does that.  It may do so specifically in a containerized environment, where environment variables can be fixed at build time, but it exposes non-containerized users to additional risk, where a malicious user can attempt to modify those variables prior to running an openssl-linked application, by pointing to a policy-violating configuration.

I would think, specifically in the situation you describe, the solution to the problem would be for the build team to construct a container that is hardened to faults by the deployment team.  If I were building such a container I would do one of the following:
1) Ensure that your openssl modules and config file implement the desired policy in the default location for openssl.  Doing this ensures, that at deployment time, with no additional effort, the container will pick up the desired policy automatically
2) If for some reason a config must be in a non-default location (say if you want to allow the selection of different config files for different use cases), the default location should contain an explicitly broken config file, which prevents the loading of openssl, and transitively the starting of the corresponding application in the container.

In short, the build team for the container, wanting to provide a secure-by-default image, should either do exactly that (by making the default configuration one that they consider secure), or prevent a successful deployment if the requirements of the container (i.e. the proper selection of deploy-time environment variables), aren't met.

**Comment by planetf1** [comment]
As per the latest comment in the PR, I will
 - work through & respond to the outstanding comments made in this issue & the PR - thanks for the detailed responses.
 
 And then, as/if appropriate
  - ensure the list of environment variables & conditions are suitably configurable

**Comment by grooverdan** [comment]
I'm having the same trouble in MariaDB packaging with CAP_IPC_LOCK while allowing users to configure via OPENSSL_CONF (ref [MDEV-33301](https://jira.mariadb.org/browse/MDEV-33301)) because of this issue.

From: https://github.com/openssl/openssl/pull/22186#issuecomment-1798115720

> **OTC: We cannot accept this PR. We do not want to hardcode exemption for specific environment variables depending on Linux-specific syscalls or Linux capability APIs. We would consider a mechanism which allows an application to specify a list of environment variable names which should not be routed via secure getenv.**

OpenSSL thankfully provides environment variables as a way to provide easy configuration for end users and system administrators to provide a rich set of functionality of OpenSSL. This ease of configuration item means that these users don't need to request application developers implement a configuration/api/interface generally for every item that openssl supports.

The impact of suid / sguid privileges of an application that can take environment variables like OPENSSL_CONF/OPENSSL_MODULES is fairly obvious, a non-priv user can run an application which because of suid can access secure keys that may have not been accessible otherwise, and in a way not envisaged by the application author/packager, expose this sensitive information to a user that didn't originally have access.

The mitigation of the writing of `secure_getenv` by glibc developers did provide a useful way of filtering this out, and in doing so they filtered out all standard capabilities, because they cannot tell which of these could be exploited.

OpenSSL developers do know exactly what the environment variables OPENSSL_CONF and OPENSSL_MODULES and all the other environment variables do, as they are all used internally by openssl. How do [capabilties(7)](https://www.man7.org/linux/man-pages/man7/capabilities.7.html) affect these (even as a non-openssl developer I should be pretty close on this):

- CAP_AUDIT_CONTROL - not at all, openssl doesn't interact with audit system
- CAP_AUDIT_READ - not at all, openssl cannot create a mulitcast list netlink socket
- CAP_AUDIT_WRITE - not at all, openssl doesn't call audit write
- CAP_BLOCK_SUSPEND - not at all, openssl doesn't call epoll (EPOLLWAKEUP)
- CAP_BPF - not at all, openssl doesn't do [e]bpf
- (skipping many due to repetition)
- CAP_DAC_OVERRIDE - dangerous to have by default, same as suid, accessing files without permission checks could expose key info
- CAP_DAC_READ_SEARCH - dangerous to have ... (as above)
- CAP_FOWNER - is a chmod in `RAND_write_file`, however surrounding code might mean no impact from this
- CAP_IPC_LOCK - mlock used in small allocations in `sh_init`, was likely to succeed otherwise. Small chance of abuse by rogue application doing lots of small allocations. (so an abused application that would OOM/swap persists a little longer as a DoS?)
- CAP_IPC_OWNER - no IPC (messages, semaphores, shm) by openssl afaik
- CAP_NET_BIND_SERVICE - could be used in `BIO_bind`, however an app has this capability because < 1024 binding are expected, the port number is in application control, and could assume running tls on these is expected
- CAP_SYS_ADMIN - might imply DAC_*, such a broad priv seems safe to exclude.

From this there are 3 main capabilities in addition to suid/sguid that provide some level of risk for the system owners that receive an application based on OpenSSL:
* CAP_DAC_OVERRIDE
* CAP_DAC_READ_SEARCH
* CAP_SYS_ADMIN

I can appreciate that OpenSSL developers don't want to get into this field, but they already have by the `secure_getenv` usage. Bypassing this by a full application implementation of OPENSSL_MODULES or OPENSSL_CONF usage without `secure_getenv`, but are we likely to get it right?

The case for an application developer, is we package up a small limited set of capabilities, not related to OpenSSL, hope the existing OPENSSL_CONF/MODULES still works. And get stuck because of `secure_getenv`. As this in glibc is determined by [getauxvec(AT_SECURE) by ld.so](https://www.man7.org/linux/man-pages/man8/ld.so.8.html) we cannot even drop capabilities before `OPENSSL_init_ssl`. The script workaround won't help with exe based `setcap 'cap_ipc_lock=+ep' /usr/sbin/mariadbd` permission that get allowed by a capability binding set. So I'm left with a runtime change (and as yet uncertain), hoping that doesn't change the AT_SECURE of getauxvec that the `secure_getenv` will use.

So please "work out of the box" is just as important as "secure out of the box". And yes, an interface to allow specific capabilities/suid/sguid with various variables would be an enhancement after this.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by grooverdan** [comment]
@t8m , the constraints on implementation - https://github.com/openssl/openssl/pull/22186#issuecomment-1798115720 provide no scope for allow listing harmless capabilities (or deny listing harmful capabilities). Does that still apply?

What alternatives exist (and would be considered?)?
* An API where an application can allow list a set of env variables that pass `secure_getenv`? (and/or deny list of env vars that must go through it).
* An API where an application provides a capability allowed (or deny) mask
* An API where the application provides its own `ossl_safe_getenv` function
* maybe if openssl doesn't want to infer the security implications of application using it just shouldn't; and `secure_getenv` gets dropped all together.

**Comment by t8m** [comment]
> What alternatives exist (and would be considered?)?
> 
>* An API where an application can allow list a set of env variables that pass `secure_getenv`? (and/or deny list of env vars that must go through it).

That sounds like something that might be acceptable. I would go by just the list of variables to skip.

>* An API where an application provides a capability allowed (or deny) mask

I do not think we want to go there.

>* An API where the application provides its own `ossl_safe_getenv` function

This might be even better than the first option.

>* maybe if openssl doesn't want to infer the security implications of application using it just shouldn't; and `secure_getenv` gets dropped all together.

Nope, I am sorry, that cannot happen.

For both the first and the third option the API should have similar constraints as the [CRYPTO_set_mem_functions()](https://docs.openssl.org/master/man3/OPENSSL_malloc/) API.
